### PR TITLE
ECS core + CMake optimizations

### PR DESCRIPTION
**Pull request by Chris-plusplus**

### CODE REVIEW

File: include/ecs/Flag.h [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#pragma once
+
+namespace arch::ecs::_details { // NOLINT
+
+/// @brief Checks if component was marked as flag-component
+template<class T>
+concept FlagComponent = T::flag_component or T::flagComponent or T::FlagComponent;
```
Comment (Tosiekdev) [inline include/ecs/Flag.h:7]: please use || (unless we agreed to using or in concepts)

Comment (Chris-plusplus) [reply include/ecs/Flag.h:7]: as far as I remember use of alternative operators (or lack thereof) [is not mandatory](https://github.com/AGH-Code-Industry/arch-org/blob/master/documents/rules/code_style.md#zach%C4%99cenie-do-u%C5%BCycia-operator%C3%B3w-alternatywnych)
I personally find logic operators to be more readable in this way

File: include/math/QDiv.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
+#pragma once
```
Comment (Tosiekdev) [inline include/math/QDiv.h:1]: Bold assumption that you know better how to optimize division that your compiler, do you have any test for that?

File: include/math/QMod.h [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#pragma once
```
Comment (Tosiekdev) [inline include/math/QMod.h:1]: In general to all math changes here: Compiler probably will do it better than you if not with -O1 then with -O3, or maybe you have test, then, please, prove me wrong

File: include/meta.h [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#pragma once
```
Comment (Tosiekdev) [inline include/meta.h:1]: Changes not related directly with ECS, I suggest putting it in separate PR

File: include/meta/Typeof.hpp [hunk 5]
```diff
@@ -6,19 +6,23 @@ namespace arch::meta::typeOf::_details {
 /// @brief Helper struct of typeof() macro
 /// @brief (does nothing)
 struct Helper {
+	template<class T>
+	requires(not std::default_initializable<T> and std::copy_constructible<T> and not std::move_constructible<T>)
```
Comment (Tosiekdev) [inline include/meta/Typeof.hpp:10]: Alternative operator, please keep it in mind

File: include/meta/rtti/Typedesc.hpp [hunk 6]
```diff
@@ -15,7 +15,7 @@ Helper::operator const T*() const noexcept {
 }
 
 template<class T>
-requires(not std::default_initializable<T>)
+requires(not std::default_initializable<T> and std::copy_constructible<T> and not std::move_constructible<T>)
```
Comment (Tosiekdev) [inline include/meta/rtti/Typedesc.hpp:18]: alternative operators

File: include/ecs/ComponentPool.hpp [hunk 7]
```diff
@@ -0,0 +1,374 @@
+#include "ComponentPool.h"
+#include <utils/Assert.h>
+
+#define TEMPLATE template<class C, class E>
+#define POOL ComponentPool<C, E>
+#define TRAITS _details::ComponentTraits<C, E>
```
Comment (Tosiekdev) [inline include/ecs/ComponentPool.hpp:6]: postfix _CE?

File: include/ecs/ComponentPool.hpp [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+
+TEMPLATE
+E& POOL::_sparseAssure(const IdT id) noexcept {
+	const size_t sparsePageNum = qdiv<ETraits::pageSize>(id);
+	// virtually equal to _sparse.resize(sparsePageNum + 1), but capacity >= sparsePageNum + 1
+	while (_sparse.size() != sparsePageNum + 1) {
+		_sparse.emplace_back();
+	}
+
+	auto&& sparsePagePtr = _sparse[sparsePageNum];
```
Comment (Tosiekdev) [inline include/ecs/ComponentPool.hpp:30]: some assertion for out of bounds?

File: include/ecs/EntityPool.hpp [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+#define TEMPLATE template<class E>
+#define POOL EntityPool<E>
+
+// https://miro.com/app/board/uXjVK4gF1DI=/?share_link_id=296698570044
+// ^ picture explanations
+
+namespace arch::ecs {
+
+TEMPLATE
+typename POOL::EntityT* POOL::_tryInitPage(const size_t n) noexcept {
```
[the remark this answers is not in the document]
Comment (Chris-plusplus) [reply include/ecs/EntityPool.hpp:15]: it was voted not to use trailing return type
https://github.com/AGH-Code-Industry/arch-org/blob/master/presentations/2024_02_20/styl_kodu.md#typ-zwracany-funkcjimetody

File: include/ecs/EntityPool.hpp [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+	// virtually equal to _sparse.resize(n + 1), but resize() makes capacity == n + 1 (bad);
+	while (_sparse.size() != n + 1) {
+		_sparse.emplace_back();
+	}
+
+	auto& page = _sparse[n];
```
[the remark this answers is not in the document]
Comment (Chris-plusplus) [reply include/ecs/EntityPool.hpp:21]: while 3 lines before makes sure of that

File: include/ecs/EntityPool.h [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+	const EntityT& _sparseGet(const IdT id) const noexcept;
+
+	SparseContainer _sparse;
+	DenseContainer _dense;
+	size_t _size = 0;
+};
+
+} // namespace arch::ecs
+
+#include "EntityPool.hpp"
```
[the remark this answers is not in the document]
Comment (Chris-plusplus) [reply include/ecs/EntityPool.h:165]: include implementation for EntityPool

File: include/ecs/EntityTraits.h [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+
+	/// @brief Id section
+	struct Id {
+		/// @brief Id type
+		using Type = IdT;
+
+		/// @brief Bit length of id
+		static inline constexpr size_t length = Specs::idLength;
+		/// @brief Bit mask of id
+		static inline constexpr EntityRawT mask = ~(std::numeric_limits<EntityRawT>::max() << length);
```
[the remark this answers is not in the document]
Comment (Chris-plusplus) [reply include/ecs/EntityTraits.h:73]: the notation ID | Version is used on examples, however code uses notation [version_bits]|[id_bits]
image

File: include/ecs/EntityTraits.hpp [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+TEMPLATE
+ENTITY::Type ENTITY::fromOthers(const Type id, const Type ver) noexcept {
+	return fromRawParts(Id::rawPart(id), Version::rawPart(ver));
+}
+
+TEMPLATE
+ENTITY::RawT ENTITY::toRaw(const Type entity) noexcept {
+	return (RawT)entity;
```
[the remark this answers is not in the document]
Comment (Chris-plusplus) [reply include/ecs/EntityTraits.hpp:29]: both are equivalent in this context

File: include/ecs/EntityTraits.hpp [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+TEMPLATE
+ID::Type ID::part(const EntityT entity) noexcept {
+	return { rawPart(entity) };
+}
+
+TEMPLATE
+TRAITS::EntityRawT ID::asRawPart(const Type id) noexcept {
+	return (EntityRawT)id & mask;
```
[the remark this answers is not in the document]
Comment (Chris-plusplus) [reply include/ecs/EntityTraits.hpp:39]: both are equivalent in this context

File: include/ecs/EntityTraits.h [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+		static inline constexpr EntityRawT nullRaw = mask;
+
+		/// @brief Obtains version part from given entity
+		static inline Type part(const EntityT entity) noexcept;
+		/// @brief Obtains version raw part from given entity
+		static inline EntityRawT asRawPart(const Type version) noexcept;
+		/// @brief Converts version part to raw part
+		static inline EntityRawT rawPart(const EntityT entity) noexcept;
+		/// @brief Checks if entity has null version
+		static inline bool hasNull(const EntityT entity) noexcept;
```
[the remark this answers is not in the document]
Comment (Chris-plusplus) [reply include/ecs/EntityTraits.h:123]: isNull would suggest that paramenter is VersionT not EntityT
null EntityT is a different concept

File: include/ecs/EntityTraits.hpp [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+TEMPLATE
+VERSION::Type VERSION::part(const EntityT entity) noexcept {
+	return Type(rawPart(entity) >> Id::length);
+}
+
+TEMPLATE
+TRAITS::EntityRawT VERSION::asRawPart(const Type version) noexcept {
+	return (EntityRawT)version << Id::length;
```
[the remark this answers is not in the document]
Comment (Chris-plusplus) [reply include/ecs/EntityTraits.hpp:76]: both are equivalent in this context

File: include/ecs/EntityTraits.hpp [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+	return { rawPart(entity) };
+}
+
+TEMPLATE
+TRAITS::EntityRawT ID::asRawPart(const Type id) noexcept {
+	return (EntityRawT)id & mask;
+}
+
+TEMPLATE
+TRAITS::EntityRawT ID::rawPart(const EntityT entity) noexcept {
```
[the remark this answers is not in the document]
Comment (Chris-plusplus) [reply include/ecs/EntityTraits.hpp:43]: idk, worked for me

File: include/ecs/InclassComponentSpecs.h [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+
+/// @brief Concept satisfied by types flaged as 'in place component' in their definition
+template<class C>
+concept AnyInclassInPlaceComponent =
+	PascalCaseInclassInPlaceComponent<C> or CamelCaseInclassInPlaceComponent<C> or SnakeCaseInclassInPlaceComponent<C>;
+
+/// @brief Concept satisfied by types that specify 'ComponentPageSize' in their definition
+template<class C>
+concept PascalCaseInclassComponentPageSize = requires {
+												 { C::ComponentPageSize } -> std::convertible_to<size_t>;
```
[the remark this answers is not in the document]
Comment (Chris-plusplus) [reply include/ecs/InclassComponentSpecs.h:28]: it is, clang-format's fault

File: include/ecs/ComponentPool.hpp (hunk 8 above)
Comment (Chris-plusplus) [reply include/ecs/ComponentPool.hpp:30]: while 3 lines before would make it obsolete

File: include/ecs/EntityPool.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (Tosiekdev) [reply include/ecs/EntityPool.h:165]: And it has no problems with compilation?

[the remark this answers is not in the document]
Comment (Chris-plusplus) [reply include/ecs/EntityPool.h:165]: no, why would it?

File: include/ecs/EntityPool.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (Tosiekdev) [reply include/ecs/EntityPool.hpp:21]: I missed it, thanks

File: include/math/QDiv.h (hunk 2 above)
Comment (Chris-plusplus) [reply include/math/QDiv.h:1]: It is how compilers optimize division/modulo by powers of 2
https://godbolt.org/z/Wo7694ezj

File: include/ecs/InclassComponentSpecs.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (Chris-plusplus) [reply include/ecs/InclassComponentSpecs.h:28]: changing 
RequiresExpressionIndentation: Keyword
to
RequiresExpressionIndentation: OuterScope
could fix that

File: include/ecs/EntityPool.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (Tosiekdev) [reply include/ecs/EntityPool.h:165]: circular includes but i guess pragma takes care of it, so please ignore it

File: include/ecs/EntityTraits.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (Tosiekdev) [reply include/ecs/EntityTraits.h:73]: If it isn't mentioned anywhere, please write that down somewhere

File: include/ecs/EntityTraits.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (Tosiekdev) [reply include/ecs/EntityTraits.hpp:29]: I know, but static cast is easier to debug and brings good practice, but you can ignore it if you want

File: include/ecs/EntityTraits.hpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (Tosiekdev) [reply include/ecs/EntityTraits.hpp:43]: Ok, we will belive it works

File: include/ecs/EntityTraits.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (Chris-plusplus) [reply include/ecs/EntityTraits.h:73]: It was on Miro board, but just mentioned tbf. Thanks 4 pointing it out

File: include/math/QDiv.h (hunk 2 above)
Comment (Tosiekdev) [reply include/math/QDiv.h:1]: Key word: compilers optimize, I think we don't need it in our code, compiler will do it better

**Review by Tosiekdev — APPROVED** [review]
Great work!

**Pull request merged**